daft/udf/row_wise.py

```diff
@@ -61,16 +63,50 @@ def __call__(self, *args: Any, **kwargs: Any) -> Expression | T:
             return Expression._row_wise_udf(self.name, self._inner, self.return_dtype, (args, kwargs), expr_args)
 
 
+def call_async_batch_with_evaluated_exprs(
+    fn: Callable[..., Awaitable[Any]],
+    return_dtype: PyDataType,
+    original_args: tuple[tuple[Any, ...], dict[str, Any]],
+    evaluated_args_list: list[list[Any]],
+) -> PySeries:
+    import asyncio
+
+    args, kwargs = original_args
+
+    tasks = []
+    for evaluated_args in evaluated_args_list:
+        new_args = [evaluated_args.pop(0) if isinstance(arg, Expression) else arg for arg in args]
+        new_kwargs = {
+            key: (evaluated_args.pop(0) if isinstance(arg, Expression) else arg) for key, arg in kwargs.items()
+        }
+        coroutine = fn(*new_args, **new_kwargs)
+        tasks.append(coroutine)
+
+    async def run_tasks() -> list[Any]:
+        return await asyncio.gather(*tasks)
+
+    dtype = DataType._from_pydatatype(return_dtype)
+
+    try:
+        # try to use existing event loop
+        event_loop = asyncio.get_running_loop()
+        outputs = asyncio.run_coroutine_threadsafe(run_tasks(), event_loop).result()
+    except RuntimeError:
+        outputs = asyncio.run(run_tasks())
```

Comment:
this seemed like a pretty reasonable initial approach to me. I think as an optimization, we could create a global event loop that we reuse instead of potentially creating a new loop every time an async udf is called

Reply:
Makes sense to me, I was also thinking of using the Tokio event loop and making `eval_expression_list` async to avoid blocking entirely.